collectors/python.d.plugin/openldap/openldap.chart.py

```diff
@@ -0,0 +1,204 @@
+            ['compare_operations', 'compare', 'incremental']
+        ]
+    },
+    'waiters': {
+        'options': [None, 'Waiters', 'waiters/s', 'ldap', 'openldap.waiters', 'line'],
+        'lines': [
+            ['write_waiters', 'write', 'incremental'],
+            ['read_waiters', 'read', 'incremental']
+        ]
+    },
+}
+
+# Stuff to gather - make tuples of DN dn and attrib to get
+SEARCH_LIST = {
+    'total_connections': (
+        'cn=Total,cn=Connections,cn=Monitor', 'monitorCounter',
+    ),
+    'bytes_sent': (
+        'cn=Bytes,cn=Statistics,cn=Monitor', 'monitorCounter',
+    ),
+    'completed_operations': (
+        'cn=Operations,cn=Monitor', 'monitorOpCompleted',
+    ),
+    'initiated_operations': (
+        'cn=Operations,cn=Monitor', 'monitorOpInitiated',
+    ),
+    'referrals_sent': (
+        'cn=Referrals,cn=Statistics,cn=Monitor', 'monitorCounter',
+    ),
+    'entries_sent': (
+        'cn=Entries,cn=Statistics,cn=Monitor', 'monitorCounter',
+    ),
+    'bind_operations': (
+        'cn=Bind,cn=Operations,cn=Monitor', 'monitorOpCompleted',
+    ),
+    'unbind_operations': (
+        'cn=Unbind,cn=Operations,cn=Monitor', 'monitorOpCompleted',
+    ),
+    'add_operations': (
+        'cn=Add,cn=Operations,cn=Monitor', 'monitorOpInitiated',
+    ),
+    'delete_operations':  (
+        'cn=Delete,cn=Operations,cn=Monitor', 'monitorOpCompleted',
+    ),
+    'modify_operations': (
+        'cn=Modify,cn=Operations,cn=Monitor', 'monitorOpCompleted',
+    ),
+    'compare_operations': (
+        'cn=Compare,cn=Operations,cn=Monitor', 'monitorOpCompleted',
+    ),
+    'search_operations': (
+        'cn=Search,cn=Operations,cn=Monitor', 'monitorOpCompleted',
+    ),
+    'write_waiters': (
+        'cn=Write,cn=Waiters,cn=Monitor', 'monitorCounter',
+    ),
+    'read_waiters': (
+        'cn=Read,cn=Waiters,cn=Monitor', 'monitorCounter',
+    ),
+}
+
+
+class Service(SimpleService):
+    def __init__(self, configuration=None, name=None):
+        SimpleService.__init__(self, configuration=configuration, name=name)
+        self.order = ORDER
+        self.definitions = CHARTS
+
+        self.server = configuration.get('server', DEFAULT_SERVER)
+        self.port = configuration.get('port', DEFAULT_PORT)
+        self.username = configuration.get('username', None)
+        self.password = configuration.get('password', None)
+        self.timeout = configuration.get('timeout', DEFAULT_TIMEOUT)
+
+        self.alive = False
+        self.conn = None
+
+    def disconnect(self):
+        if self.conn:
+            self.conn.unbind()
+            self.conn = None
+            self.alive = False
+
+    def connect(self):
+        try:
+            self.conn = ldap.initialize('ldap://%s:%s' % (self.server, self.port))
+            self.conn.set_option(ldap.OPT_NETWORK_TIMEOUT, self.timeout)
+            if self.username and self.password:
+                self.conn.simple_bind(self.username, self.password)
```

Comment:
if `self.conn.simple_bind` doesn't work with `None, None` please add `if not (self.username and self.password)` to `check()`.

Reply:
> Ok after some test, seems that simple bind method doesn't like 'None' values.

Can you show a traceback? (`python.d.plugin openldap debug trace`)

Reply:
here it is:
```
2018-11-14 09:35:34: python.d INFO: plugin: main: Using python 2
2018-11-14 09:35:34: python.d ERROR: openldap: openldap: check() unhandled exception: must be string, not None
2018-11-14 09:35:34: python.d ERROR: openldap: openldap: Traceback (most recent call last):
  File "/usr/libexec/netdata/plugins.d/python.d.plugin", line 277, in check_job
    check_ok = bool(job.check())
  File "/usr/libexec/netdata/python.d/openldap.chart.py", line 183, in check
    return self.connect() and self.get_data()
  File "/usr/libexec/netdata/python.d/openldap.chart.py", line 165, in connect
    self.conn.simple_bind(self.username, self.password)
  File "/usr/lib64/python2.7/site-packages/ldap/ldapobject.py", line 201, in simple_bind
    return self._ldap_call(self._l.simple_bind,who,cred,RequestControlTuples(serverctrls),RequestControlTuples(clientctrls))
  File "/usr/lib64/python2.7/site-packages/ldap/ldapobject.py", line 99, in _ldap_call
    result = func(*args,**kwargs)
TypeError: must be string, not None

2018-11-14 09:35:34: python.d ERROR: openldap: openldap: check() => [FAILED]
2018-11-14 09:35:34: python.d ERROR: openldap: openldap: Traceback (most recent call last):
  File "/usr/libexec/netdata/plugins.d/python.d.plugin", line 19, in <module>
    from time import monotonic as time
ImportError: cannot import name monotonic

2018-11-14 09:35:34: python.d INFO: plugin: main: FINISHED
```

Reply:
> if self.conn.simple_bind doesn't work with None, None please add if not (self.username and self.password) to check().

But we don't even use `conn.simple_bind` in case username and password are not provided.

Reply:
Wait, we don't need `conn.simple_bind` for anonymous connection?

Reply:
>  num = self.conn.search(dn, ldap.SCOPE_BASE, 'objectClass=*', [attr, ])
result_type, result_data = self.conn.result(num, 1)

this ^^ works without `simple_bind` ?

Reply:
If I add this check, in case there is an anonymous bind the check will never allow the connection to be established.  Anonymous bind doesn't even need a simple_bind method to be called.

That's why we use simple_bind ONLY if username and password are provided and are NOT None:

```
if self.username and self.password:
                self.conn.simple_bind(self.username, self.password)
```

Is there something wrong in my thinking?

Reply:
> Anonymous bind doesn't even need a simple_bind method to be called.

I was not aware of this. If so, then yes, your code is OK.